### PR TITLE
Adding VM's allocation status column to the machines table

### PR DESCRIPTION
**Pull request by anshulvermapatel**
### Which issue this PR addresses:
Fixes
https://msazure.visualstudio.com/AzureRedHatOpenShift/_boards/board/t/Firefly/Stories/?workitem=9777520

### What this PR does / why we need it:
Adds a column to the Machines table which displays the allocation status of the corresponding VM.

### CODE REVIEW

File: pkg/portal/cluster/machines.go
Comment (cadenmarchese) [inline pkg/portal/cluster/machines.go:97]: Should we leave out this `if` statement since we're already using the `resourceType eq 'Microsoft.Compute/virtualMachines` filter in the call to `ListByResourceGroup` above?

File: pkg/portal/cluster/fetcher.go
Comment (cadenmarchese) [inline pkg/portal/cluster/fetcher.go:68]: Maybe:
```suggestion
func newRealFetcher(log *logrus.Entry, dialer proxy.Dialer, doc *api.OpenShiftClusterDocument, fetcher azureSideFetcher) (*realFetcher, error) {
```

File: pkg/portal/cluster/machines_test.go
Comment (devamanv) [inline pkg/portal/cluster/machines_test.go:130]:
```suggestion
		NewResourceClientFunction = features.NewResourcesClient
		NewVirtualMachineClientFunction = compute.NewVirtualMachinesClient
```
Both assignments should be reversed.

File: pkg/portal/cluster/machines_test.go
Comment (devamanv) [inline pkg/portal/cluster/machines_test.go:140]: The test name could be made a little more descriptive and should clearly state the functionality you're trying to test.

File: pkg/portal/cluster/fetcher.go
Comment (devamanv) [inline pkg/portal/cluster/fetcher.go:25]: Since we have made changes to the interface, can we check to see if it requires generating and committing new  mocks for the interface?

File: pkg/portal/cluster/machines.go
Comment (devamanv) [inline pkg/portal/cluster/machines.go:18]: The intent here is to not make them local(which they aren't anyway, as they are exported), but to be able to swap the function definitions or make them swappable during testing. So, consider changing the comment.

File: pkg/portal/cluster/machines.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines.go:97]: I think yes .. I am not able to recall .. why I put that condition there .. looking at the code now, it looks like the `if` condition is not required, hence removing it.

File: cmd/aro/portal.go
Comment (anshulvermapatel) [reply cmd/aro/portal.go:29]: done

File: pkg/portal/cluster/fetcher.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/fetcher.go:68]: So there is another interface called fetcher .. so I don't want someone to confuse that fetcher with this one. This struct is specifically to fetch something from Azure side .. anything in future we want to fetch from Azure, this struct can be used or can be enhanced to fetch more stuff may be.

File: pkg/portal/cluster/machines_test.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines_test.go:130]: Ohh yes .. my bad .. made the changes.

File: pkg/portal/cluster/machines_test.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines_test.go:140]: Yup .. changed the title, do check and let me know.

File: pkg/portal/cluster/fetcher.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/fetcher.go:25]: We do not have mocks for fetcher right now. Maybe we can generate them in a separate work item.

File: pkg/portal/cluster/machines.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines.go:18]: So here .. in the comment, I have talked about creating local variables not making the functions local. Yes they are exported .. which is not required so I will make them local and also will change the comment to make it more descriptive to give an idea of why that is done.

File: cmd/aro/portal.go
Comment (facchettos) [reply cmd/aro/portal.go:29]: no :D

Comment (anshulvermapatel) [reply cmd/aro/portal.go]: pushed again with removed comments. This time for sure.

File: pkg/portal/cluster.go
Comment (facchettos) [inline pkg/portal/cluster.go:191]: can we check the error here ?

File: pkg/portal/portal.go
Comment (facchettos) [reply pkg/portal/portal.go:314]: also, why do we use path prefix here? shouldn't it be just `Path`?

File: pkg/portal/security_test.go
Comment (facchettos) [inline pkg/portal/security_test.go:52]: comment

File: pkg/portal/info_test.go
Comment (facchettos) [inline pkg/portal/info_test.go:26]: comment

File: pkg/portal/cluster/machines.go
Comment (facchettos) [inline pkg/portal/cluster/machines.go:109]: maybe doesn't matter too much, but isn't the real condition `string.HasPrefix(*status.Code, "PowerState/")` ?

File: pkg/portal/cluster/fetcher.go
Comment (facchettos) [inline pkg/portal/cluster/fetcher.go:58]: I am not sure I understand the point of this, it is not really a fetcher, and its fields could just be in the `realFetcher`

**Review by facchettos — CHANGES_REQUESTED** [review]

File: pkg/portal/cluster/machines.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines.go:21]: Since these are the directly exported functions from a package, there were not getting mocked with a function with desired return value hence I need to resort to this.

File: pkg/portal/cluster.go
Comment (anshulvermapatel) [reply pkg/portal/cluster.go:191]: Yes, .. I have updated the code and will push it with other changes.

File: pkg/portal/cluster/machines.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines.go:72]: I thought of making it a type to make sure whoever reads my code knows what this map is for.

File: pkg/portal/cluster/machines.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines.go:114]: Whats wrong with continue here? Is this for performance optimization or readability optimization?

Comment (anshulvermapatel) [reply pkg/portal/cluster/machines.go:114]: PowerState is at least some state as it comes in the for of "PowerState/{somestate}" for example "PowerState/running" and I take only the status code which has "PowerState" in it. 
Since it is looping through the VMs .. there is no chance of a VMs powerstate getting overridden.

File: pkg/portal/security_test.go
Comment (anshulvermapatel) [reply pkg/portal/security_test.go:52]: done

File: pkg/portal/portal.go
Comment (anshulvermapatel) [reply pkg/portal/portal.go:314]: No, the comment is not necessary.
I just used path prefix because it was used before. I am changing this for all the other routes as well

File: pkg/portal/info_test.go
Comment (anshulvermapatel) [reply pkg/portal/info_test.go:26]: done

File: pkg/portal/cluster/machines.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines.go:109]: this condition can also be used, however, I have seen the statuses, they have powerstate only for actual power status of the VM.

File: pkg/portal/cluster/fetcher.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/fetcher.go:58]: So the realFetcher is something which has stuff used for fetching details about a cluster which is in cosmosDB. I wanted to open doors to fetch from Azure directly like VM status VMID etc and things required for that can land in this struct. If any other stuff in future requires to fetch anything from Azure which might require any other information for that, it can be added here.

File: pkg/portal/cluster/machines.go
Comment (facchettos) [reply pkg/portal/cluster/machines.go]: it breaks the flow, so it's more readability related. and also most `continue` are actually just a sign that you should split your code into smaller functions

File: pkg/portal/cluster/machines.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines.go:21]: Okay .. but wherever I pass the interface as parameter to the `VMAllocationStatus` method from, there I will have to call these functions to get the interface object and I will have to anyways test that as well .. which will put me in the same situation ..

File: pkg/portal/cluster/machines.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines.go:72]: Yeah okay .. I too now think that .. I will change it ..

File: pkg/portal/cluster/machines.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines.go:106]: if the convention changes .. we will have to change the code anyway ..

File: pkg/portal/cluster/fetcher.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/fetcher.go:58]: but then we will do it in a segregated manner which I think will help a new engineer to understand the code easily.

File: pkg/portal/cluster/machines.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines.go]: Are you saying to put it like this -
```
func putAllocationStatusToMap(ctx context.Context, clusterRGName string, vmAllocationStatus map[string]string, res mgmtfeatures.GenericResourceExpanded, virtualMachineClient compute.VirtualMachinesClient, log *logrus.Entry) {
	var vmName, allocationStatus string
	vm, err := virtualMachineClient.Get(ctx, clusterRGName, *res.Name, mgmtcompute.InstanceView)
	if err != nil {
		log.Warn(err) // can happen when the ARM cache is lagging
		return
	}
	vmName = *vm.Name
	instanceViewStatuses := vm.InstanceView.Statuses
	for _, status := range *instanceViewStatuses {
		if strings.Contains(*status.Code, "PowerState") {
			allocationStatus = *status.Code
		}
	}
	vmAllocationStatus[vmName] = allocationStatus
}
```

File: pkg/portal/cluster/fetcher.go
Comment (facchettos) [reply pkg/portal/cluster/fetcher.go:58]: this does not help right now, it's not even a fetcher so it's confusing. I stand by my comment

File: pkg/portal/cluster/machines.go
Comment (facchettos) [reply pkg/portal/cluster/machines.go:21]: Having the interfaces in the struct is a standard pattern (passing it as a parameter is also possible but imo having these as a field is better here), and with what you have done it means that we can modify the functions at runtime and this is dangerous (package wide function variable is not a good idea)

File: pkg/portal/cluster/machines.go
Comment (facchettos) [reply pkg/portal/cluster/machines.go:106]: in any case, one is more correct than the other so I think we should go for the more precise one

Comment (anshulvermapatel) [reply pkg/portal/cluster/machines.go:106]: Okay .. I can do that ..

File: pkg/portal/cluster/fetcher.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/fetcher.go:58]: so what design do you suggest here?

File: pkg/portal/cluster/machines.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines.go]: I have added the above function and removed the continue line.

File: pkg/portal/cluster/machines.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines.go:21]: in the tests .. I am resetting these variable to the original functions in the defer statement so that any other tests which might use those variables have the original function from the `features` package.

File: pkg/portal/cluster/machines.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines.go]: done, resolving this conversation

File: pkg/portal/cluster/machines.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines.go]: done, resolving

File: pkg/portal/cluster/machines.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines.go]: Changed, resolving conversation.

File: pkg/portal/cluster/machines.go
Comment (facchettos) [reply pkg/portal/cluster/machines.go:23]: use an interface.

File: pkg/portal/cluster/fetcher.go
Comment (facchettos) [reply pkg/portal/cluster/fetcher.go:58]: I already suggested a way to do it

Comment (anshulvermapatel) [reply pkg/portal/cluster/fetcher.go:58]: to put the fields in the realfetcher itself .. I really think a segregation must be here to the required things to fetch from Azure(not from kube-apiserver).. maybe we can call it something else as you mentioned it is not a fetcher .. we can call it something else but I think the segregation must be there.

Comment (facchettos) [reply pkg/portal/cluster/fetcher.go:68]: actually it should be a completely separated client. we don't need any of the existing realfetcher underlying clients do we? (only log ) 
the real fetcher fetches stuff from kubernetes and it should keep doing just that. If we need to fetch from azure, we should create a new and separate client

**Review by facchettos — CHANGES_REQUESTED** [review]
I think the azurefetcher should be separated from the realfetcher as it actually is something else entirely and we don't need to extend the existing realfetcher which only deals with the kube api with azure capabilities

File: pkg/portal/cluster/machines_test.go
Comment (facchettos) [inline pkg/portal/cluster/machines_test.go:182]: setting up the mock is better inside the test itself, not in the test data. if it proves difficult to generalize your test, that probably means that it should be in different test functions

File: pkg/portal/cluster/machines.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines.go:72]: we use it elsewhere, we use it in the `portal` package.

File: pkg/portal/cluster/fetcher.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/fetcher.go:29]: Okay, I have made the changes, will pushed it after making more changes as per the review.

File: pkg/portal/cluster/machines_test.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines_test.go:182]: I should depend on the test case right .. the data we want and test our function according to that data. That should come in the test data only.

File: pkg/portal/cluster/machines.go
Comment (facchettos) [reply pkg/portal/cluster/machines.go:111]: then remove that line from my suggestion and keep the rest of it

Comment (anshulvermapatel) [reply pkg/portal/cluster/machines.go:111]: ohh yaa, sorry, if we do not get any status from then it is going to be empty string, I did this to keep the UI clean as this is going to get seen in the UI as is.

Comment (anshulvermapatel) [reply pkg/portal/cluster/machines.go:111]: I just want to have all the VMNames as keys in the map.

File: pkg/portal/cluster/machines.go
Comment (facchettos) [reply pkg/portal/cluster/machines.go:72]: well one of these seems to be exported for no reason. Because it portal we apparently use the fetchclient interface, not the realfetcher itself.

File: pkg/portal/cluster/machines.go
Comment (facchettos) [reply pkg/portal/cluster/machines.go:111]: then change `"unknown"` to `""`

Comment (anshulvermapatel) [reply pkg/portal/cluster/machines.go:111]: No, all the values of the map will have empty string, I have declared the variable `allocationStatus` at the starting of the function, if there is value for the vm allocationStatus code, the var will have that value else it would be empty string only. So I think it is good as is.

File: pkg/portal/cluster/machines.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines.go:72]: Yes, it is fetchClient, I think I made it capital because VM is a short form and wrote it in caps I think. I can make it in small letter.

File: pkg/portal/cluster/machines_test.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines_test.go:182]: Sorry, I might have misunderstood you. 
So what I think is, it should depend on the test, support if the test is about a situation where the function will return an error in the first few lines itself, so we would not expect the later function/methods to get executed and we won't write <mockInterface>.EXPECT().<function>.whatever. 

If you meant something else than what I understood, please do elaborate a little.

File: pkg/portal/cluster/machines.go
Comment (facchettos) [reply pkg/portal/cluster/machines.go:111]: take a look again at my suggestion, it does exactly that. the `return` statement break the flow

File: pkg/portal/cluster/machines_test.go
Comment (facchettos) [reply pkg/portal/cluster/machines_test.go:182]: but you should do that setup in the test itself, if there is no expected call you should know it in the data, and set it up accordingly.

File: pkg/portal/cluster/machines.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines.go:111]: Yes yes .. you are right .. I am changing it to empty String.

File: pkg/portal/cluster/machines_test.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines_test.go:182]: I have seen this as standard structure in our codebase only .. check this - https://github.com/Azure/ARO-RP/blob/782894071cf91e8c1ca6e182417b851925f823cc/pkg/api/validate/dynamic/dynamic_test.go#L43-L110

Comment (facchettos) [reply pkg/portal/cluster/machines_test.go:182]: I know, jim and mj loved that , but it creates loads of code duplication, unreadable and overly complicated tests and we should stop using it

**Review by ellis-johnson — COMMENTED** [review]
Changes look awesome! Just two things

Firstly it seems like the machines state isn't getting updated when you switch between different machines. It just stays on the first one you selected regardless of which one you select next. I tried this for the other InfoLists we use for Nodes and MachineSets and it seems to work fine so I think this is an isolated issue

Secondly it would be awesome if we can get an e2e test for the Machines tab from this

File: pkg/portal/cluster/machines_test.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines_test.go]: I have updated this in the latest push, please do check and let me know.

**Review by ellis-johnson — CHANGES_REQUESTED** [review]
Looks good! Just need to rebuild the portal and commit the files and it will be all good to go

File: pkg/portal/cluster/machines_test.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines_test.go]: Done please do check

File: pkg/portal/cluster/machines_test.go
Comment (facchettos) [inline pkg/portal/cluster/machines_test.go:240]: left over comments

**Review by facchettos — DISMISSED** [review]
something to fix in a follow up but otherwise LGTM (conditional to what ellis requested being implemented)

File: pkg/portal/cluster/machines_test.go
Comment (anshulvermapatel) [reply pkg/portal/cluster/machines_test.go]: removed these leftover comments.
And yes the bug which Ellis pointed out, I fixed it in this PR.

**Review by ellis-johnson — APPROVED** [review]
LGTM!

**Review by facchettos — APPROVED** [review]
please make sure to squash before merging (whoever merges)

**Pull request merged**
